keps/sig-testing/3041-node-conformance-and-features/README.md

```diff
@@ -0,0 +1,803 @@
+
+- There is no way to disable tests for a specific feature gate
+  (example where it’s needed: https://github.com/kubernetes/kubernetes/issues/99854).
+  It is a "feature" guarded by feature gate, but not marked as "feature".
+- There are no tests validating that k8s works with ALL beta feature gates disabled.
+  Creating such test would be hard with today's test labeling.
+- Some tests are degraded in how `[Feature]` and `[NodeFeature]` labels are applied
+  (for example: https://github.com/kubernetes/kubernetes/pull/105921).
+- labels that we apply heavily depend on the environment tests are running on.
+  For example, https://github.com/kubernetes/kubernetes/pull/104803 is not marked
+  as `Feature`, even though it depends on the environment - if `test-handler`
+  is not installed, test will not succeed. Similar thing to AppArmor tests.
+
+### Goals
+
+Split the special environment and feature development labels:
+
+- Clarify the semantic of a `[Feature:]` label.
+- Introduce the `[FeatureGate:]` label with the stage (`[Alpha]`, `[Beta]`, etc.)
+  addition, replace tags like `[NodeAlphaFeature]`.
+
+Clean up labels that means whether special environment is needed:
+
+- Eliminate the `[NodeFeature:]` label by renaming to `[Feature:]` when applicable.
+- Document the meaning of `[NodeConformance]` label.
+- Remove the `[NodeSpecialFeature:]` label.
+- Introduce the `[Environment:Foo]` label to indicate the special environment
+  that is not a standard OSS test infrastructure machines needed to run the feature.
+
+### Non-Goals
+
+- As `[NodeConformance]` may be a confusing term and may be confused with the
+  k8s Conformance tests, renaming of it is not in a scope of this KEP.
+- Convert textual tags into the ginkgo v2 tags. This is optional, but not
+  required and can be handled outside of this KEP.
+
+## Proposal
+
+### Clarify the semantic of a `[Feature:]` label
+
+The current [feature definition is](https://github.com/kubernetes/community/blob/32a1c14d04ff78684d78b827ac7c49f70352d509/contributors/devel/sig-testing/e2e-tests.md#kinds-of-tests):
+
+> * `[Feature:.+]`: If a test has non-default requirements to run or targets
+>   some non-core functionality, and thus should not be run as part of the
+>   standard suite, it receives a `[Feature:.+]` label, e.g.
+>   `[Feature:Performance]` or `[Feature:Ingress]`. `[Feature:.+]` tests are not
+>   run in our core suites, instead running in custom suites. If a feature is
+>   experimental or alpha and is not enabled by default due to being incomplete
+>   or potentially subject to breaking changes, it does not block PR merges, and
+>   thus should run in some separate test suites owned by the feature owner(s)
+>   (see Continuous Integration below).
+
+Feature will indicate that things may NOT work on all k8s distros and/or node
+capabilities. This flag will not indicate the Feature Gate state.
+
+Proposed change:
+
+> * `[Feature:.+]`: If a test validates functionality that may only work
```

Comment:
It's worth noting that renaming Feature has massive churn requirement on CI configs upstream and downstream. The default state of most jobs is to opt out of features.

Reply:
yes, the back compat was mostly the reason to keep the `Feature` naming. Alternative may be `Capability`, but again, it will require a lot of renaming.

Reply:
Makes sense. Perhaps add that rationale to the KEP?